### PR TITLE
implement authorities and leaves sub commands

### PR DESCRIPTION
**Pull request by joshua-rutherford**
In addition to reviewing the code here, I'm looking for review of the command tree.

acert authorities create - create an authority identity
acert authorities list - list the authority identities
acert authorities delete FINGERPRINT - delete an authority identity
acert authorities issue FINGERPRINT - issue a leaf identity from an authority identity

acert leaves list - list the leaf identities
acert leaves delete FINGERPRINT - delete a leaf identity
acert leaves export FINGERPRINT - export a leaf identity

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,6 +20,78 @@ To print the current version and commit for the acert binary run the following c
...
+
+To list the authorities run the following command:
+
+    acert authorities list
+
+### Leaves
+
+Leaves represent the X.509 identities of users or services and cannot issue identities.
+
+#### Creating
```
Comment (justincely) [inline README.md]: The command is `issue`, so i think the heading should match

File: README.md [hunk 2]
```diff
@@ -20,6 +20,78 @@ To print the current version and commit for the acert binary run the following c
...
+
+    acert authorities delete FINGERPRINT
+
+#### Listing
+
+To list the authorities run the following command:
+
+    acert authorities list
+
+### Leaves
```
Comment (justincely) [inline README.md]: i get why you called it a leaf, but is this common terminology for this? I feel like `acert identities export <whatever>` matches my experience more - but i'll defer that to your experience with x.509

File: cmd/authorities/create/options.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"time"
+)
+
+// Options defines the options for the create command.
+type Options struct {
+
+	// CommonName defines the common name for an authority.
+	CommonName string `mapstrucutre:"commonName"`
```
Comment (justincely) [inline cmd/authorities/create/options.go]: mispelled `mapstructure` throughout this

File: cmd/authorities/issue/options.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	"time"
+)
+
+// Options defines the options for the issue command.
+type Options struct {
+
+	// CommonName defines the common name for an certificate.
+	CommonName string `mapstrucutre:"commonName"`
```
Comment (justincely) [inline cmd/authorities/issue/options.go]: more `mapstrucutre`

File: cmd/leaves/export/command.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+			if err != nil {
+				return err
+			}
+
+			leaf, err := leaves.Fetch(args[0])
+			if err != nil {
+				return err
+			}
+
+			switch strings.ToLower(options.Type) {
```
Comment (justincely) [inline cmd/leaves/export/command.go]: no checking if the string does not match those 3.  If i spell wrong, or input gibberish, it does nothing silently. 

```
justin@simon:~/codebase/acert$ acert leaves export 8bf7e24781c3 -f pem -t sdaf
justin@simon:~/codebase/acert$
```

File: cmd/leaves/export/options.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package export
+
+// Options defines the options for the export command.
+type Options struct {
+
+	// Type defines the type of artifact to export.
+	Type string `mapstrucutre:"type"`
```
Comment (justincely) [inline cmd/leaves/export/options.go]: sp

File: README.md (hunk 2 above)
Comment (joshua-rutherford) [reply README.md]: The issue is that an authority is an identity as well.  The only difference is it's an identity that can sign other identities.  I agree that it is a bit less common though accepted (https://en.wikipedia.org/wiki/Public_key_certificate#End-entity_or_leaf_certificate).  I do not want to make it identity though as that's ambiguous in a very real sense.  Other proposals?

File: cmd/authorities/create/options.go (hunk 3 above)
Comment (joshua-rutherford) [reply cmd/authorities/create/options.go]: fixed.

File: cmd/leaves/export/command.go (hunk 5 above)
Comment (joshua-rutherford) [reply cmd/leaves/export/command.go]: Good catch, fixed to:
```
$ ./bin/acert leaves export 38c63470363e -f pem -t sdaf 
Error: error parsing type [sdaf] must be one of [authority, certificate, key]
Usage:
  acert leaves export FINGERPRINT [flags]

Flags:
  -f, --format string   the format of the exported leaf (default "pem")
  -h, --help            help for export
  -t, --type string     the type of values to be exported [authority, certificate, key] (default "certificate")

error parsing type [sdaf] must be one of [authority, certificate, key]
```

File: cmd/leaves/export/options.go (hunk 6 above)
Comment (joshua-rutherford) [reply cmd/leaves/export/options.go]: fixed.

File: README.md (hunk 2 above)
Comment (justincely) [reply README.md]: nothing I can think of is very universal. I'm good with leaf, since it's at least somewhat common/documented.

**Pull request merged**